torch/_functorch/aot_autograd.py

```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
             # Case where none of the aliases have a ._base
             # we generate a synthetic base without gradients, and generate views off of it
             example_idx = aliased_input_indices[0]
-            synthetic_base = torch.Tensor(fwd_inputs[example_idx].untyped_storage())
+            if is_runtime:
+                # Wondering when we hit this case? a known example is when compiling
+                # the optimizer step, where the parameters came from an LSTM module.
+                # LSTM performs weight packing, so weights are views from a shared buffer.
+                # NOTE: There is one case where this is unsafe:
+                # torch.Tensor(storage) will ALWAYS create a 1D tensor, which is not necessarily
+                # the same shape as the "actual" base that the tensor came from.
+                # For the most part this is fine, because we always use as_strided()
+                # to generate the original aliased inputs again.
+                # If we were to use view-replay though, this could cause the aliased views
+                # to have incorrect sizes.
+                example_alias = fwd_inputs[example_idx]
+                # We don't actually have a convenient way of going from storage -> tensor,
+                # So using set_() here (we suffer some minor overhead, but this case is rare).
+                synthetic_base = torch.empty((), dtype=example_alias.dtype, device=example_alias.device)
+                synthetic_base.set_(example_alias.untyped_storage())
+            else:
+                # Construct a FakeTensor synthetic base, given any of our FakeTensor aliases.
+                # (Assuming that all our aliases share the same FakeTensorMode, which should
+                # generally be true)
+                curr_aliased_inp = fwd_inputs[example_idx]
+                # We need to hardcode that the synthetic base we're creating here is a FakeTensor.
+                # We should also be guaranteed that torch_dispatch subclasses have been desugared
+                # by the time we get here (through AOTDispatch).
+                assert isinstance(curr_aliased_inp, FakeTensor)
+                curr_fake_mode = curr_aliased_inp.fake_mode
+                # Hack: we can't construct a FakeTensor directly from a storage,
+                # so just plumb the storage through a temporary real tensor.
+                synthetic_base_dummy = torch.Tensor(fwd_inputs[example_idx].untyped_storage())
+                synthetic_base_ = curr_fake_mode.from_tensor(synthetic_base_dummy)
+                synthetic_base = synthetic_base_.to(device=curr_aliased_inp.device)
```

Comment:
The way we create a fake tensor with a specific storage in meta_utils.py is by allocating a fake tensor, and then set_'ing the fake storage into it. It's not clear why you can't do that here. In fact, this would permit you to use precisely the code in the `is_runtime` branch.

Reply:
Thanks - I played around with doing this for a bit, and realized the situation is even better. When this function is run under tracing, `synthetic_base = torch.empty((0,), dtype=example_alias.dtype, device=example_alias.device)` which just do the right thing already and create a FakeTensor, since our entire trace function is run under FakeTensorMode.